31-Password Generator/script.js

```diff
@@ -0,0 +1,65 @@
+const resultEl = document.getElementById('result')
+const lengthEl = document.getElementById('length')
+const uppercaseEl = document.getElementById('uppercase')
+const lowercaseEl = document.getElementById('lowercase')
+const numbersEl = document.getElementById('numbers')
+const symbolsEl = document.getElementById('symbols')
+const generateEl = document.getElementById('generate')
+const clipboardEl = document.getElementById('clipboard')
+
+const randomFunc = {
+  lower: getRandomLower,
+  upper: getRandomUpper,
+  number: getRandomNumber,
+  symbol: getRandomSymbol
+}
+
+clipboardEl.addEventListener("click", () => {
+  const password = resultEl.innerText;
+  navigator.clipboard.writeText(password);
+  alert('Password Copied To Clipboard')
+});
+
+generateEl.addEventListener('click', () => {
+  const length = +lengthEl.value
+  const hasLower = lowercaseEl.checked
+  const hasUpper = uppercaseEl.checked
+  const hasNumber = numbersEl.checked
+  const hasSymbol = symbolsEl.checked
+
+  resultEl.innerText = generatePassword(hasLower, hasUpper, hasNumber, hasSymbol, length)
+})
+
+function generatePassword(lower, upper, number, symbol, length) {
+  let generatedPassword = ''
+  const typesCount = lower + upper + number + symbol
+  const typesArr = [{ lower }, { upper }, { number }, { symbol }].filter(item => Object.values(item)[0])
+  if (typesCount === 0) { return '' }
+
+  for (let i = 0; i < length; i += typesCount) {
+    typesArr.forEach(type => {
+      const funcName = Object.keys(type)[0]
+      generatedPassword += randomFunc[funcName]()
+    })
+  }
+
+  const finalPassword = generatedPassword.slice(0, length)
+  return finalPassword
+}
+
+function getRandomLower() {
+  return String.fromCharCode(Math.floor(Math.random() * 26) + 97)
+}
+
+function getRandomUpper() {
+  return String.fromCharCode(Math.floor(Math.random() * 26) + 65)
+}
+
+function getRandomNumber() {
+  return String.fromCharCode(Math.floor(Math.random() * 10) + 48)
+}
```

Comment:
Suggestion, all the numbers above seem like magic numbers to me, we spoke of the problem of magic numbers before, I do understand that you are referring to the ASCII table but that doesn't mean that everyone who looks at your code knows that!
here is my suggestion, some where before all theses functions we go ahead and create a new constants as follows
```
const ASCII_CAPITAL_A = 65;
const ASCII_SMALL_A = 97;
const ASCII_ZERO = 48;
const ALPHABET_LENGTH = 26;
const SINGE_DIGITS_LENGTH = 10;
```

then we use them as follows
```
function getRandomLower() {
  return String.fromCharCode(Math.floor(Math.random() * ALPHABET_LENGTH) + ASCII_SMALL_A)
}
...
...
```

Reply:
bro i wish all the programers on earth would be more careful with how every one will look on their code ,
and make sure  they understand it (specially brad ) ,   the same  way you do with mine😂😂

Problem Solved.